OpenRA.Mods.Common/Traits/World/Locomotor.cs

```diff
@@ -280,31 +351,143 @@ bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor, CellConditions
 			}
 
 			// If we cannot crush the other actor in our way, we are blocked.
-			if (Crushes.IsEmpty)
+			if (Info.Crushes.IsEmpty)
 				return true;
 
 			// If the other actor in our way cannot be crushed, we are blocked.
 			// PERF: Avoid LINQ.
 			var crushables = otherActor.TraitsImplementing<ICrushable>();
 			foreach (var crushable in crushables)
-				if (crushable.CrushableBy(otherActor, self, Crushes))
+				if (crushable.CrushableBy(otherActor, self, Info.Crushes))
 					return false;
 
 			return true;
 		}
 
-		public virtual bool DisableDomainPassabilityCheck { get { return false; } }
+		static bool IsMovingInMyDirection(Actor self, Actor other)
+		{
+			// PERF: Because we can be sure that OccupiesSpace is Mobile here we can save some performance by avoiding querying for the trait.
+			var otherMobile = other.OccupiesSpace as Mobile;
+			if (otherMobile == null || !otherMobile.CurrentMovementTypes.HasMovementType(MovementType.Horizontal))
+				return false;
 
-		public virtual object Create(ActorInitializer init) { return new Locomotor(init.Self, this); }
-	}
+			// PERF: Same here.
+			var selfMobile = self.OccupiesSpace as Mobile;
+			if (selfMobile == null)
+				return false;
 
-	public class Locomotor
-	{
-		public readonly LocomotorInfo Info;
+			// Moving in the same direction if the facing delta is between +/- 90 degrees
+			var delta = Util.NormalizeFacing(otherMobile.Facing - selfMobile.Facing);
+			return delta < 64 || delta > 192;
+		}
 
-		public Locomotor(Actor self, LocomotorInfo info)
+		public void WorldLoaded(World w, WorldRenderer wr)
 		{
-			Info = info;
+			world = w;
+			var map = w.Map;
+			actorMap = w.ActorMap;
+			actorMap.CellsUpdated += CellsUpdated;
+			pathabilityCache = new CellLayer<CellCache>(map);
+
+			terrainInfos = Info.TilesetTerrainInfo[map.Rules.TileSet];
+
+			allPlayerMask = world.AllPlayerMask;
+
+			foreach (var cell in map.AllCells)
+				UpdateCellCost(cell);
+
+			map.CustomTerrain.CellEntryChanged += MapCellEntryChanged;
+			map.Tiles.CellEntryChanged += MapCellEntryChanged;
+		}
+
+		void CellsUpdated(IEnumerable<CPos> cells)
```

Comment:
The logic here is a bit inconsistent: `UpdateCellBlocking` runs on an individual cell, and is looped here, while `UpdateTerrainCells` runs on all cells and does its own looping internally.

Reply:
This was just an ugly workaround for when i removed the ITick from locomotor. Either we could update the cell cost directly in MapCellEntryChanged or Add the ITick again to Locomotor for updating all the cells that the terrain has changed.